constraintlayout/constraintlayout/src/main/java/androidx/constraintlayout/helper/widget/CircularFlow.java

```diff
@@ -0,0 +1,252 @@
+    }
+
+
+    public float[] getAngles() {
+        return Arrays.copyOf(mAngles, mCountAngle);
+    }
+
+
+    @Override
+    protected void init(AttributeSet attrs) {
+        super.init(attrs);
+        if (attrs != null) {
+            TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.ConstraintLayout_Layout);
+            final int N = a.getIndexCount();
+
+            for (int i = 0; i < N; i++) {
+                int attr = a.getIndex(i);
+                if (attr == R.styleable.ConstraintLayout_Layout_circularflow_viewCenter) {
+                    mViewCenter = a.getResourceId(attr, 0);
+                } else if (attr == R.styleable.ConstraintLayout_Layout_circularflow_angles) {
+                    mReferenceAngles = a.getString(attr);
+                    setAngles(mReferenceAngles);
+                } else if (attr == R.styleable.ConstraintLayout_Layout_circularflow_radiusInDP) {
+                    mReferenceRadius = a.getString(attr);
+                    setRadius(mReferenceRadius);
+                }
+            }
+            a.recycle();
+        }
+    }
+
+    @Override
+    public void onAttachedToWindow() {
+        super.onAttachedToWindow();
+        if (mReferenceAngles != null) {
+            setAngles(mReferenceAngles);
+        }
+        if (mReferenceRadius != null) {
+            setRadius(mReferenceRadius);
+        }
+        anchorReferences();
+    }
+
+    private void anchorReferences() {
+        mContainer = (ConstraintLayout) getParent();
+        ConstraintSet c = new ConstraintSet();
+        c.clone(mContainer);
+        for (int i = 0; i <= mCount; i++) {
+            int id = mIds[i];
+            View view = mContainer.getViewById(id);
+
+            if (view != null) {
+                int radius = DEFAULT_RADIUS;
+                float angle = DEFAULT_ANGLE;
+
+                if (i < getRadius().length){
+                    radius = getRadius()[i];
+                } else {
+                    Log.e("CircularFlow", "Added radius to view with id: " + mMap.get(view.getId()));
+                }
+
+                if (i < getAngles().length){
+                    angle = getAngles()[i];
+                } else {
+                    Log.e("CircularFlow", "Added angle to view with id: " + mMap.get(view.getId()));
+                }
+                c.constrainCircle(view.getId(), mViewCenter, radius, angle);
+            }
+        }
+        c.applyTo(mContainer);
+        applyLayoutFeatures();
+    }
+
+    /**
+     * Add a view to the CircularFlow. The referenced view need to be a child of the container parent.
+     * The view also need to have its id set in order to be added.
+     * The views previous need to have its radius and angle set in order to be added correctly a new view.
+     * @param view
+     * @param radius
+     * @param angle
+     * @return
+     */
+    public void addViewToCircularFlow(View view, int radius, float angle) {
```

Comment:
should you have the corresponding removeView() ? or is the one in ConstraintHelper doing enough?

Reply:
I added a `removeViewFromCircularFlow` where remove view with `removeView` from ConstraintHelper, and remove radius/angles for this view.